### PR TITLE
SchemaAwareJsonDataProvider improvements

### PR DESCRIPTION
**Pull request by feelgood-interface**
1. Use ActualSchema for array item, so all references are resolved.
2. Check for nullable properties.
3. Use the schema information to set RootTableName and ArrayValueName.

### CODE REVIEW

**Comment by jbartlau** [comment]
@feelgood-interface Thanks for the PR - we'll review the code and get back here. Appreciated very much!

File: combit.ListLabel.SchemaAwareJsonDataProvider/JsonSchemaTable/JsonSchemaOnlyTableRow.cs [hunk 1]
```diff
@@ -33,31 +33,34 @@ internal static JsonTableColumn ColumnFromSchemaData(string columnName, JsonSche
...
-                return new JsonTableColumn(columnName, typeof(bool), null);
+                return new JsonTableColumn(columnName, isNullable ? typeof(bool?) : typeof(bool), null);
             }
 
             //https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7.3.1
-            if (data.Type == JsonObjectType.String && (data.Format == "date-time" || data.Format == "date" || data.Format == "time"))
+            if (data.Type.HasFlag(JsonObjectType.String) && (data.Format == "date-time" || data.Format == "date" || data.Format == "time"))
             {
-                return new JsonTableColumn(columnName, typeof(DateTime), null);
+                return new JsonTableColumn(columnName, isNullable ? typeof(DateTime?) : typeof(DateTime), null);
```
Comment (jbartlau) [inline combit.ListLabel.SchemaAwareJsonDataProvider/JsonSchemaTable/JsonSchemaOnlyTableRow.cs]: Actually, on the List & Label side all types are nullable by default. Thus, whether a type is passed in as nullable or not shouldn't change the behavior. Can you elaborate a bit on why this change is required and why it's not added for the string type below? For symmetrical reasons I would have expected it there as well 🙂.

File: combit.ListLabel.SchemaAwareJsonDataProvider/SchemaAwareJsonDataProvider.cs [hunk 2]
```diff
@@ -35,6 +38,12 @@ public string Schema
                 _schemaLocation = value;
                 var fileProvider = SchemaFileProvider ?? Options?.FileProvider ?? new NetworkFileProvider();
                 _schema = JsonSchema.FromJsonAsync(fileProvider.ReadAsString(_schemaLocation)).Result;
+
+                RootTableName = _schema.Title ?? RootTableName;
```
Comment (jbartlau) [inline combit.ListLabel.SchemaAwareJsonDataProvider/SchemaAwareJsonDataProvider.cs:42]: This may break existing projects that rely on the "old" logic of retrieving the table name. However, as this provider is mostly used in source code form and the addition of a compatibility switch would be trivial, I support this change although it might be breaking.

File: combit.ListLabel.SchemaAwareJsonDataProvider/SchemaAwareJsonDataProvider.cs [hunk 3]
```diff
@@ -35,6 +38,12 @@ public string Schema
...
+                if (Data.IsArray)
+                {
+                    ArrayValueName = _schema.Definitions.FirstOrDefault().Key ?? ArrayValueName;
```
Comment (jbartlau) [inline combit.ListLabel.SchemaAwareJsonDataProvider/SchemaAwareJsonDataProvider.cs]: Not sure as schematized JSON is not my primary science 😁 - could FirstOrDefault() also return `null` here? In that case, it might be better to use `_schema.Definitions.FirstOrDefault()?.Key ?? ArrayValueName`?

**Review by jbartlau — COMMENTED** [review]
Thanks again @feelgood-interface - I did a quick review, maybe you could quickly comment on my questions. I'll merge the PR then once we are on the same page.

File: combit.ListLabel.SchemaAwareJsonDataProvider/JsonSchemaTable/JsonSchemaOnlyTableRow.cs (hunk 1 above)
Comment (feelgood-interface) [reply combit.ListLabel.SchemaAwareJsonDataProvider/JsonSchemaTable/JsonSchemaOnlyTableRow.cs]: If L&L types are null by default, this is not required. String is a reference type, so it can be null and older C# versions (<8) doesn't support nullable reference types.

File: combit.ListLabel.SchemaAwareJsonDataProvider/SchemaAwareJsonDataProvider.cs (hunk 2 above)
Comment (feelgood-interface) [reply combit.ListLabel.SchemaAwareJsonDataProvider/SchemaAwareJsonDataProvider.cs:42]: I added this, because one cannot set the RootTableName from JsonSchema info without loading the JsonSchema from outside. So maybe we can make the JsonSchema property public and then remove the new RootTableName and ArrayValueName setters.

File: combit.ListLabel.SchemaAwareJsonDataProvider/JsonSchemaTable/JsonSchemaOnlyTableRow.cs (hunk 1 above)
Comment (jbartlau) [reply combit.ListLabel.SchemaAwareJsonDataProvider/JsonSchemaTable/JsonSchemaOnlyTableRow.cs]: Yup, they are definitely nullable - you can pass NULL for any of our types out of the box. In this case, you could consider to push a change here?

File: combit.ListLabel.SchemaAwareJsonDataProvider/SchemaAwareJsonDataProvider.cs (hunk 3 above)
Comment (jbartlau) [reply combit.ListLabel.SchemaAwareJsonDataProvider/SchemaAwareJsonDataProvider.cs]: Resolved by your commit [588ee9a](https://github.com/combit/NETDataProviders/commit/588ee9a41eff5706bde68487874f4600eef0f04f)

File: combit.ListLabel.SchemaAwareJsonDataProvider/SchemaAwareJsonDataProvider.cs (hunk 2 above)
Comment (jbartlau) [reply combit.ListLabel.SchemaAwareJsonDataProvider/SchemaAwareJsonDataProvider.cs:42]: I think your implementation has advantages in keeping the data schema information where it belongs. And if someone should have a problem with this change, they may easily add a compat switch to keep the "old" behavior.

File: combit.ListLabel.SchemaAwareJsonDataProvider/SchemaAwareJsonDataProvider.cs (hunk 3 above)
Comment (feelgood-interface) [reply combit.ListLabel.SchemaAwareJsonDataProvider/SchemaAwareJsonDataProvider.cs]: `IDictionary<string, JsonSchema>.FirstOrDefault()` will return a default `KeyValuePair<string, JsonSchema>` with `string` (i.e. the key) defaulting to `null`. However, `Definitions` can be null, so I added a null check there.

File: combit.ListLabel.SchemaAwareJsonDataProvider/JsonSchemaTable/JsonSchemaOnlyTableRow.cs (hunk 1 above)
Comment (jbartlau) [reply combit.ListLabel.SchemaAwareJsonDataProvider/JsonSchemaTable/JsonSchemaOnlyTableRow.cs]: Resolved by your commit [b00a99a](https://github.com/combit/NETDataProviders/commit/b00a99a1a43df98dbe1e3a77f13ed24ce24d37ab)

**Comment by jbartlau** [comment]
@feelgood-interface Thanks - this has been merged.
